Spike: Async instead of EM::Synchrony

ColinDKelley opened the pull request:
- Use Ruby 3.2.1.
- Replace `EM` and `EM::Synchrony` with `Async`.
- Replace raw `Fiber` with `Async::Task` (which wraps `Fiber`).
- Replace special fibered `Mutex`, `ConditionVariable` and `Monitor` with regular Ruby 3.2.1 implementations.
- Drop support for Rails 4 and 5. Assert that Rails version is 6.

Note: diff will look cleaner with Hide Whitespace checked: https://github.com/Invoca/fibered_mysql2/pull/9/files?diff=unified&w=1

ColinDKelley commented on fibered_mysql2.gemspec:
```diff
@@ -29,6 +29,6 @@ Gem::Specification.new do |spec|
   spec.require_paths = ["lib"]
 
-  spec.add_dependency 'em-synchrony', '~> 1.0'
+  spec.add_dependency 'async'
```
@invocandg : I'm wondering about renaming the gem here to something like `async_mysql2`. That would allow it to coexist with the previous version of `fibered_mysql2`. Well, almost. I'd also need to change the module name, so they wouldn't collide when required.

ColinDKelley replied:
I tried that and pushed it, but github decided to show this PR as if every file was deleted and replaced. So I'll make it a separate PR.

ColinDKelley replied:
As mentioned in Slack, I got that working, but I'm backing away since it doesn't seem necessary now that we're walking away from the `USE_FALCON` env var and just make a pure-Falcon/Async branch.

invocandg commented on lib/active_record/connection_adapters/fibered_mysql2_adapter.rb:
```diff
@@ -1,82 +1,72 @@
-      if in_use?
-        # Because we are actively releasing connections from dead fibers, we only want
-        # to enforce that we're expiring the current fibers connection, iff the owner
+      if (ot = owner_task)
```
Looking at the `owner_task` as defined below

```
def owner_task
  @owner.nil? || @owner.is_a?(Async::Task) or
    raise "@owner must be an Async::Task! Found #{@owner.inspect}"
   
  @owner
end
```

`@owner` could be `nil` isn't it? In which case all the subsequent calls on it like `alive?` could fail isn't it?

So we may need to add another check for not nil like `!ot.nil? && ot.alive? ...` ?

ColinDKelley replied:
But short-circuiting `||` means that if it's `nil`, we won't check `alive?`, right?

ColinDKelley replied:
Also as mentioned in Slack, I'm leaning back to using `Fiber` as the owning object. Even though `Async::Task` seems more precise and less subject to corner-cases in Ruby... but as soon as we move to Rails 7, we're going to be using `isolation_level: :fiber`, so we might as well now.

ColinDKelley replied:
^ Done.

invocandg commented on lib/active_record/connection_adapters/fibered_mysql2_adapter.rb:
```diff
@@ -1,82 +1,72 @@
         @owner = nil
       else
-        raise ::ActiveRecord::ActiveRecordError, "Cannot expire connection, it is not currently leased."
+        raise ::ActiveRecord::ActiveRecordError, "Cannot expire connection; it is not currently leased."
```
Shouldn't `ActiveRecord::ConnectionNotEstablished` more appropriate error here?

ColinDKelley replied:
This code was adapted [straight from Rails](https://github.com/rails/rails/blob/99cd347a55fd9bddb458afd003e0f50c5b2ba29a/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L254), so I wanted to keep the exception signatures the same.

invocandg commented on lib/active_record/connection_adapters/fibered_mysql2_adapter.rb:
```diff
@@ -1,82 +1,72 @@
         end
       else
-        raise ::ActiveRecord::ActiveRecordError, "Cannot steal connection, it is not currently leased."
+        raise ::ActiveRecord::ActiveRecordError, "Cannot steal connection; it is not currently leased."
```
Same as above -

Shouldn't ActiveRecord::ConnectionNotEstablished more appropriate error here?

ColinDKelley replied:
Same [here](https://github.com/rails/rails/blob/99cd347a55fd9bddb458afd003e0f50c5b2ba29a/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L267).

invocandg commented on spec/unit/async_task_spec.rb:
```diff
@@ -0,0 +1,31 @@
+
+require 'fibered_mysql2/async_task'
+
+RSpec.describe FiberedMysql2::AsyncTask do
```
👍

invocandg reviewed (approved):
LGTM apart from minor feedback and clarification question.

ColinDKelley commented:
@invocandg Can you review the changes I pushed since you last reviewed?

dcaddell commented on lib/fibered_mysql2/fibered_database_connection_pool.rb:
```diff
@@ -2,239 +2,35 @@
-      @mon_count += 1
-      true
+  module FiberedDatabaseConnectionPool
+    case ::Rails::VERSION::MAJOR
+    when 6
+    else
+      raise ArgumentError, "unexpected Rails version #{Rails::VERSION::MAJOR}"
```
Are we trying to preserve this `case` structure for supporting Rails 7+ in the future? It's a little awkward while Rails 6 is the only supported major version.

Speaking of: can the gemspec be updated to advertise this contract?

ColinDKelley replied:
Good points! Fixed: https://github.com/Invoca/fibered_mysql2/pull/9/commits/87996bc250f0a812f4bd3e9b9ff80245745a1666

invocandg commented on lib/active_record/connection_adapters/fibered_mysql2_adapter.rb:
```diff
@@ -1,30 +1,19 @@
 module FiberedMysql2
-  module FiberedMysql2Adapter_4_2
+  module FiberedMysql2Adapter_6
     def lease
```
(nit) It looks a lot like https://github.com/rails/rails/blob/main/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L298-L365 except the use of `Fiber.current` over `Thread.current`.

Any particular reason why we cannot just adapt as is from the rails source code, which certainly feels a bit more readable with the use of `in_use?`

ColinDKelley replied:
I started there years ago.

But these methods that look at `in_use?` followed by `@owner` kept tripping me up as I studied them, because `in_use?` is just an alias that returns `@owner`, and asking for the owner twice actually creates a risk of a race condition.

invocandg commented on spec/unit/fibered_mysql2_adapter_spec.rb:
```diff
@@ -28,19 +28,25 @@
-    it { should eq(Fiber.current) }
+    it { in_concurrent_environment { should eq(Fiber.current) } }
 
     if Rails::VERSION::MAJOR > 4
```
You mean 6?

ColinDKelley replied:
Yeah, those tests are unnecessary now, so I got rid of them: https://github.com/Invoca/fibered_mysql2/pull/9/commits/2d2f46502cba52d179255670ba6dfd68d25de034

invocandg reviewed (approved):
LGTM, apart from minor feedback

invocandg commented on spec/unit/fibered_database_connection_pool_spec.rb:
```diff
@@ -378,99 +329,99 @@ def cancel_timer(timer_block)
           connection_pool.checkin(c0)
-          em_helper.run_next_ticks
+          task1.wait
+          while task1_fiber.alive?
```
>Too bad there's no Fiber#wait. 

+1

invocandg commented:
@ColinDKelley I am not sure how clean you want to get in this PR considering this started out as a spike. 

Considering this spike turned out to be an actual implementation along with the tests and docs, there are few areas that could be addressed/updated if you are up for it:

- `FiberedMysql2 adds Fiber support to `ActiveRecord::ConnectionAdapters::EMMysql2Adapter`` at https://github.com/Invoca/fibered_mysql2/blob/master/README.md?plain=1#L5

- `Behaves the same as `ActiveRecord::ConnectionAdapters::EMMysql2Adapter`  at https://github.com/Invoca/fibered_mysql2/blob/master/README.md?plain=1#L28
 
- `Tested with Rails versions 5.2 and 6.0.` at https://github.com/Invoca/fibered_mysql2/blob/master/README.md?plain=1#L24

- I think we can drop `https://github.com/Invoca/fibered_mysql2/blob/spike/async-instead-of-synchrony/lib/fibered_mysql2/fibered_mutex_with_waiter_priority.rb` as there is no more EM related code here

invocandg commented on lib/fibered_mysql2.rb:
```diff
@@ -1,5 +1,6 @@
 # frozen_string_literal: true
 
+require 'async'
 require 'fibered_mysql2/version'
```
(nit) Its more cleaner to have it as follows:

```
require_relative 'active_record/connection_adapters/fibered_mysql2_adapter'
require_relative 'fibered_mysql2/fibered_database_connection_pool'
require_relative 'fibered_mysql2/fibered_mutex_with_waiter_priority'
require_relative 'fibered_mysql2/fibered_mysql2_connection_factory'
require_relative 'fibered_mysql2/version'
```

ColinDKelley replied:
Agreed! Fixed: https://github.com/Invoca/fibered_mysql2/pull/9/commits/910b8bc095daf3136e7a2c167035d72f2ea9c663

ColinDKelley commented:
> Considering this spike turned out to be an actual implementation along with the tests and docs, there are few areas that could be addressed/updated if you are up for it:

> FiberedMysql2 adds Fiber support to ActiveRecord::ConnectionAdapters::EMMysql2Adapter at https://github.com/Invoca/fibered_mysql2/blob/master/README.md?plain=1#L5

> Behaves the same as ActiveRecord::ConnectionAdapters::EMMysql2Adapter at https://github.com/Invoca/fibered_mysql2/blob/master/README.md?plain=1#L28

> Tested with Rails versions 5.2 and 6.0. at https://github.com/Invoca/fibered_mysql2/blob/master/README.md?plain=1#L24

Fixed, along with some other README cleanup: https://github.com/Invoca/fibered_mysql2/pull/9/commits/756060880a98ac7e9f6691d053ffc3febbffbea0

> I think we can drop https://github.com/Invoca/fibered_mysql2/blob/spike/async-instead-of-synchrony/lib/fibered_mysql2/fibered_mutex_with_waiter_priority.rb as there is no more EM related code here

Right! Fixed: https://github.com/Invoca/fibered_mysql2/pull/9/commits/a56dcf9cf15941c26b8b3b4352b77e313c7da1af

I also got `bin/console` working again, including removing the dependency on `Rails::VERSION` and instead using `ActiveRecord::VERSION`: https://github.com/Invoca/fibered_mysql2/pull/9/commits/3905aac77e1177087251fc6b1057b97520d8787e

invocandg commented on README.md:
```diff
@@ -2,7 +2,8 @@
 # FiberedMysql2
 
-FiberedMysql2 adds Fiber support to `ActiveRecord::ConnectionAdapters::EMMysql2Adapter`
+FiberedMysql2 adds Fiber support to `ActiveRecord::ConnectionAdapters::Mysql2Adapter` for Rails 6.
```
👍
